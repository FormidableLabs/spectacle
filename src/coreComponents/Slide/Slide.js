import React from 'react';
import PropTypes from 'prop-types';

import useSlide, { SlideContext } from '../../hooks/useSlide';
import { DeckContext } from '../../hooks/useDeck';
import isComponentType from '../../utils/isComponentType.js';
import { SlideContainer } from './Slide.style';

/**
 * Slide component wraps anything going in a slide and maintains
 * the slides' internal state through useSlide.
 */

export const Slide = ({ children, slideNum }) => {
  const [, , , keyboardControls] = React.useContext(DeckContext);
  const initialState = { currentSlideElement: 0, immediate: false };

  const numberOfSlideElements = Array.isArray(children)
    ? children.filter(x => isComponentType(x, 'SlideElementWrapper')).length
    : 0;

  return (
    <SlideContainer>
      <SlideContext.Provider
        value={useSlide(
          initialState,
          slideNum,
          numberOfSlideElements,
          keyboardControls
        )}
      >
        {children}
      </SlideContext.Provider>
    </SlideContainer>
  );
};

Slide.propTypes = {
  children: PropTypes.node.isRequired,
  slideNum: PropTypes.number.isRequired
};
