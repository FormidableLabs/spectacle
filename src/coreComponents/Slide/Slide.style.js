import styled from 'styled-components';

const SlideContainer = styled.div`
  overflow-x: hidden;
  height: inherit;
  width: inherit;
`;

export { SlideContainer };
