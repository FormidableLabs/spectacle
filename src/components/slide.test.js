import React from 'react';
import { mount } from 'enzyme';
import Slide from './slide';
import Appear from './appear';

const _mockContext = function() {
  return {
    styles: {
      global: {
        body: {
          background: '#eee'
        }
      },
      components: {
        content: {}
      }
    },
    store: {
      getState: () => ({ route: { params: '', slide: 0 } }),
      subscribe: () => {},
      dispatch: () => {}
    }
  };
};

describe('<Slide />', () => {
  beforeEach(() => {
    window.watchMedia = jest.fn();
    window.matchMedia = jest.fn().mockReturnValue({ matches: [] });
  });

  afterEach(() => {
    window.watchMedia = null;
    window.matchMedia = null;
  });

  test('should render correctly without transitions.', () => {
    const wrapper = mount(
      <Slide>
        <div>Slide Content</div>
      </Slide>,
      { context: _mockContext() }
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render correctly with transitions.', () => {
    const wrapper = mount(
      <Slide transition={['slide', 'spin']}>
        <div>Slide Content</div>
      </Slide>,
      { context: _mockContext() }
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should return the correct transition keys', () => {
    const wrapper = mount(
      <Slide transitionIn={['slide']} transitionOut={['fade']}>
        <div>Slide Content</div>
      </Slide>,
      { context: _mockContext() }
    );

    expect(wrapper.instance().getTransitionKeys()).toEqual(['slide']);
    wrapper.setState({ reverse: true });
    expect(wrapper.instance().getTransitionKeys()).toEqual(['fade']);
  });

  test('should call optional callback when slide becomes active', () => {
    const spy = jest.fn();
    mount(
      <Slide onActive={spy} slideIndex={5}>
        <div>Slide Content</div>
      </Slide>,
      { context: _mockContext() }
    );
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toBeCalledWith(5);
  });

  test('should create <Appear /> fragments with their appearance in order', () => {
    const spy = jest.fn();
    mount(
      <Slide slideIndex={4} dispatch={spy} hash={4}>
        <Appear order={2} fragment={{ fragments: [] }}>
          <div className="second">This shows second</div>
        </Appear>
        <Appear order={3} fragment={{ fragments: [] }}>
          <div className="third">This shows third</div>
        </Appear>
        <Appear order={1} fragment={{ fragments: [] }}>
          <div className="first">This shows first</div>
        </Appear>
      </Slide>,
      { context: _mockContext() }
    );
    expect(spy).toHaveBeenCalledTimes(3);
    expect(spy.mock.calls).toEqual([
      [
        {
          payload: {
            slide: 4,
            id: '4-0',
            animations: [false],
            className: 'fragment first'
          },
          type: 'ADD_FRAGMENT'
        }
      ],
      [
        {
          payload: {
            slide: 4,
            id: '4-1',
            animations: [false],
            className: 'fragment second'
          },
          type: 'ADD_FRAGMENT'
        }
      ],
      [
        {
          payload: {
            slide: 4,
            id: '4-2',
            animations: [false],
            className: 'fragment third'
          },
          type: 'ADD_FRAGMENT'
        }
      ]
    ]);
  });

  test('should order <Appear /> fragments without an order first', () => {
    const spy = jest.fn();
    mount(
      <Slide slideIndex={7} dispatch={spy} hash={7}>
        <Appear order={1} fragment={{ fragments: [] }}>
          <div className="first">This shows second</div>
        </Appear>
        <Appear fragment={{ fragments: [] }}>
          <div className="no-order">This shows third</div>
        </Appear>
        <Appear order={2} fragment={{ fragments: [] }}>
          <div className="second">This shows first</div>
        </Appear>
      </Slide>,
      { context: _mockContext() }
    );
    expect(spy).toHaveBeenCalledTimes(3);
    expect(spy.mock.calls).toEqual([
      [
        {
          payload: {
            slide: 7,
            id: '7-0',
            animations: [false],
            className: 'fragment no-order'
          },
          type: 'ADD_FRAGMENT'
        }
      ],
      [
        {
          payload: {
            slide: 7,
            id: '7-1',
            animations: [false],
            className: 'fragment first'
          },
          type: 'ADD_FRAGMENT'
        }
      ],
      [
        {
          payload: {
            slide: 7,
            id: '7-2',
            animations: [false],
            className: 'fragment second'
          },
          type: 'ADD_FRAGMENT'
        }
      ]
    ]);
  });
});
