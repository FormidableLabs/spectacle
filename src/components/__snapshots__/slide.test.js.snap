// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slide /> should render correctly with transitions. 1`] = `
<Slide
  align="center center"
  presenterStyle={Object {}}
  style={Object {}}
  transition={
    Array [
      "slide",
      "spin",
    ]
  }
  viewerScaleMode={false}
>
  <VictoryAnimation
    data={
      Object {
        "transform": " translate3d(-100%, 0, 0) rotateY(-90deg)",
        "zIndex": 1,
      }
    }
    delay={0}
    duration={1000}
    easing="quadInOut"
  >
    <div
      className="spectacle-slide"
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#eee",
          "display": "flex",
          "height": "100%",
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "top": 0,
          "transform": " translate3d(-100%, 0, 0) rotateY(-90deg)",
          "width": "100%",
          "zIndex": 1,
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "position": "relative",
          }
        }
      >
        <div
          className=" spectacle-content"
          data-radium={true}
          style={
            Object {
              "flex": 1,
              "maxHeight": 700,
              "maxWidth": 1000,
              "padding": 10,
              "transform": "scale(1)",
            }
          }
        >
          <div>
            Slide Content
          </div>
        </div>
      </div>
    </div>
  </VictoryAnimation>
</Slide>
`;

exports[`<Slide /> should render correctly without transitions. 1`] = `
<Slide
  align="center center"
  presenterStyle={Object {}}
  style={Object {}}
  viewerScaleMode={false}
>
  <VictoryAnimation
    data={
      Object {
        "transform": " translate3d(0px, 0px, 0px)",
        "zIndex": 1,
      }
    }
    delay={0}
    duration={1000}
    easing="quadInOut"
  >
    <div
      className="spectacle-slide"
      data-radium={true}
      style={
        Object {
          "backgroundColor": "#eee",
          "display": "flex",
          "height": "100%",
          "left": 0,
          "overflow": "hidden",
          "position": "absolute",
          "top": 0,
          "transform": " translate3d(0px, 0px, 0px)",
          "width": "100%",
          "zIndex": 1,
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "position": "relative",
          }
        }
      >
        <div
          className=" spectacle-content"
          data-radium={true}
          style={
            Object {
              "flex": 1,
              "maxHeight": 700,
              "maxWidth": 1000,
              "padding": 10,
              "transform": "scale(1)",
            }
          }
        >
          <div>
            Slide Content
          </div>
        </div>
      </div>
    </div>
  </VictoryAnimation>
</Slide>
`;
